### PR TITLE
[receiver/mongodbatlas] Support collection of MongoDBAtlas Access Logs

### PR DESCRIPTION
**Pull request by dehaansa**
**Description:** Add the ability to collect MongoDB Atlas Access Logs

**Link to tracking Issue:** #21182

**Testing:** Unit test coverage & integration test based on a recently retrieved payload.

**Documentation:** Added documentation of configuration to the Readme.

### CODE REVIEW

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (schmikei) [inline receiver/mongodbatlasreceiver/access_logs.go:211]: I can imagine this being a bottleneck for logs. Say we are getting the max number of log entries of 20000 and we keep adding logs to the fullLogs. We only ever send the logs down at once so we could theoretically stream logs down the telemetry pipeline. Could we try consuming logs after each request when safe to not overload processors with large batches of logs?

File: receiver/mongodbatlasreceiver/config.go
Comment (schmikei) [inline receiver/mongodbatlasreceiver/config.go:94]: Do you have thoughts on having a maximum number of access logs we will go back to i.e. something similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a4e4fe4af4f7c7fe0deb8720c92db1b5801461a6/receiver/mongodbatlasreceiver/config.go#L74

where we give users a way to configure a hard limit of the number of logs we will go back into the history for?

**Review by schmikei — CHANGES_REQUESTED** [review]
Overall looks pretty good!! I have a couple concerns about upper limits and giving users control over a heavily accessed mongodbatlas and how this scales. But overall nice work!! 

Some really nice refactors around the project config loading I love to see 🚀

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/access_logs.go:272]: Since I made all of the other fields into log attributes, I kept just the logline as the body, but I'm certainly open to changing that.

File: receiver/mongodbatlasreceiver/config.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/config.go:94]: Seems reasonable.

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (schmikei) [reply receiver/mongodbatlasreceiver/access_logs.go:286]: Seems mostly reasonable, I'll probably defer to Dan's opinion when he gets back, since I think he's more knowledgeable about this than I am :)

File: receiver/mongodbatlasreceiver/README.md
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/README.md:78]: It's a bit confusing to me that access logs are a separate section from other logs. Is this really necessary? It seems to introduce a lot of unnecessary duplication.

We already include audit logs as an optional type of logs. Wouldn't it make sense to include access logs there too?

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (djaglowski) [reply receiver/mongodbatlasreceiver/access_logs.go:286]: If we've captured all the other information as attributes, then I think it's appropriate to include the message as a string body.

**Review by djaglowski — COMMENTED** [review]
This looks good to me except I wonder if the separate configuration structure is appropriate.

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/access_logs.go:286]: The "message" is actually just another (stringified) JSON object. See example below. All of the other attributes are promoted to attributes, and then 'logline' is parsed as json and set as the body. I don't think the logLine is consistent enough in content that it makes sense to try to parse that and promote all of its attributes, too.

```
{
        "authResult": false,
        "authSource": "admin",
        "failureReason": "UserNotFound: User \"mms-automation@admin\" not found",
        "groupId": "5bc762b579358e3332046e6a",
        "hostname": "host1.mongodb.net",
        "ipAddress": "555.555.555.555",
        "logLine": "{\"t\":{\"$date\":\"2023-04-26T02:38:56.454+00:00\"},\"s\":\"I\",  \"c\":\"ACCESS\",   \"id\":20249,   \"ctx\":\"conn371\",\"msg\":\"Authentication failed\",\"attr\":{\"mechanism\":\"SCRAM-SHA-1\",\"speculative\":true,\"principalName\":\"mms-automation\",\"authenticationDatabase\":\"admin\",\"remote\":\"555.555.555.555:56444\",\"extraInfo\":{},\"error\":\"UserNotFound: User \\\"mms-automation@admin\\\" not found\"}}",
        "timestamp": "Wed Apr 26 02:38:56 GMT 2023",
        "username": "mms-automation"
    },
    ```

File: receiver/mongodbatlasreceiver/README.md
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/README.md:78]: We lose page size controls if we swap access logs to just a flag in project & embed in logs. We could then add access_log_page_size and access_log_max_pages, does that still feel like the right direction to you?

Comment (dehaansa) [reply receiver/mongodbatlasreceiver/README.md:78]: There's also the possibility that a user would want access logs, but not the rest of the logs. I don't see a way to configure that in the combined config path.

**Review by djaglowski — APPROVED** [review]

**Comment by dehaansa** [comment]
Since this got an approval - don't merge yet, figured out a timing issue in testing where logs are sometimes not available in the API for upwards of 1 minute, need to factor this delay into the start/end calculations to avoid data loss.

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (schmikei) [inline receiver/mongodbatlasreceiver/access_logs.go:76]: I feel like this is the wrong collection interval as described in the README.md this is the 

> - `collection_interval` (default `3m`) This receiver collects metrics on an interval. Valid time units are `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`.

I think there should be something in the configuration that is separate lever for users to use to determine how often we poll.

File: receiver/mongodbatlasreceiver/README.md
Comment (schmikei) [reply receiver/mongodbatlasreceiver/README.md:69]: This also leads us into questioning whether these should be on by default if a user is getting logs. I feel like we should maybe disable by default at first as to not increase costs of current users and then follow up with enabling it? 

@djaglowski @dehaansa do you guys have any thoughts?

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/access_logs.go:76]: Currently access logs are enabled per-project, so we would then have a lever pre-project to define poll interval, unless you had a different idea.

**Review by schmikei — CHANGES_REQUESTED** [review]
Just requesting changes cause I have some minor concerns/design decisions I want to get clarification on (they may be necessary, just wanted to discuss them first).

Also was curious if users would ever want to run `logs` without project `access_logs`? It doesn't seem like that's a use case of the current implementation or its not a documented case in the testdata/README.md

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (schmikei) [reply receiver/mongodbatlasreceiver/access_logs.go:76]: Yeah my biggest gripe was that the documentation didn't match the usage. Could we perhaps have a field per `LogsProjectConfig` with a reasonable default? That's at least where my head was at with how I expected this to work.

File: receiver/mongodbatlasreceiver/README.md
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/README.md:69]: My comment should be clearer. Access Log are only "enabled" by default if you provide at least `access_logs:`, so they are realistically disabled by default but that could be made much more clear.

As far as the `enabled` pattern, I also do not like it but was trying to make MongoDBAtlas internally consistent since there are so many different potential components.

File: receiver/mongodbatlasreceiver/factory.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/factory.go:92]: Sure, will refactor to check before creating.

File: receiver/mongodbatlasreceiver/README.md
Comment (djaglowski) [reply receiver/mongodbatlasreceiver/README.md:69]: > @djaglowski correct me if I'm wrong but I think we are trying to avoid the `enabled` portions of configs with new functionality and rather opt for existence of fields as to whether they are enabled? I.e trying to reduce a redundant parameter flow.

There's been some recent discussion on this and I was wrong to make this assumption as others disagree. I think it's probably best to leave it, especially since we're using it elsewhere in this receiver.

Comment (djaglowski) [reply receiver/mongodbatlasreceiver/README.md:69]: > This also leads us into questioning whether these should be on by default if a user is getting logs. I feel like we should maybe disable by default at first as to not increase costs of current users and then follow up with enabling it?
> 
> @djaglowski @dehaansa do you guys have any thoughts?

I agree with leaving it disabled by default, to protect current users. As to whether that should change in the future - probably not unless users are asking for it.

File: receiver/mongodbatlasreceiver/access_logs.go
Comment (dehaansa) [reply receiver/mongodbatlasreceiver/access_logs.go]: Refactored to allow poll interval config per project.

**Review by schmikei — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
